### ISSUE TITLE
Pickling of Bio.Align.PairwiseAligner for Multiprocessing

### ISSUE THREAD

**Issue by gcambray** [biopython/biopython#3485]
### Setup

Biopython 1.78
Python 3.7 
OSX

Hello there,

This is more a question than an issue.

I am trying to parallelize pairwise alignments using multiprocessing
For this to be efficient, I would need to avoid re-instantiating PairwiseAligner in each subprocess and so would need an instance to be pickable.

I have naively tried to extend the class to implement a __getstate__ and __setstate__ methods, only to realize that arguments are actually modifying the underlying _aligners.PairwiseAligner, which itself refers to _algorithms.PairwiseAligner in ways that I don't fully understand...

I'm completely out of my depth here, hence a very (hopefully not too) vague question 👍 is there a way to make PairwiseAligner objects pickable?

With thanks

**Comment by mdehoon** [comment]
It is possible in principle, though I haven't tried it myself and probably it would take some effort to get this working Not much in terms of programming, it's more about going through the Python documentation and figure out how pickling works under the hood.

But are you sure that this is really the bottleneck? Since you need to initialize the aligner only once for each process, I would think that this takes a very small amount of time compared to running the alignments.

**Comment by gcambray** [comment]
Thanks for the quick response.
I am using multiprocessing.pool so what actually happen in the absence of serialization is that the aligner is initialized in the function that is mapped, so with every sequence to align... I guess there would be ways to initialize once per process but that would likely involve loading all the sequences in memory first.
Anyway, I figured I can initialize a subclassed aligner first with dummy getstate and setstate method that actually do nothing and assign the aligners attributes that do not get pickled easily from within the mapped function. This seems to perform well enough.
For example, with a million sequence to align, the multiprocessed script run in 30s when the aligner is initialized with every sequence, down to 22s with only attribute assigned within the mapped function, as compared to 19s when I don't assign the attributes and get default alignment behaviors.
So it seems that pickling the aligner would gain me 3s.
Pickling would thus still make a difference with larger sequence numbers.

As a side note, it seems that substitution_matrices.Array object are still not pickable (I think I saw some tickets about this)
e.g.:
 
`import pickle
a = pickle.dumps(self.scoring_matrix)
b = pickle.loads(a)
self.scoring_matrix
Array([[ 1., -1., -1., -1.,  1.],
       [-1.,  1., -1., -1.,  1.],
       [-1., -1.,  1., -1.,  1.],
       [-1., -1., -1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.]],
         alphabet='ATGCN')
b
<Bio.Align.substitution_matrices.Array object at 0x7fe3d0781f50>
self.scoring_matrix.alphabet
'ATGCN'
b.alphabet
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/gcambray/anaconda3/envs/py3/lib/python3.7/site-packages/Bio/Align/substitution_matrices/__init__.py", line 245, in alphabet
    return self._alphabet
AttributeError: 'Array' object has no attribute '_alphabet'`

**Comment by gcambray** [comment]
Following up on this:
When pickling an extended PairwiseAligner object as follow

```
from Bio.Align import PairwiseAligner
import pickle

class PickablePairwiseAligner(PairwiseAligner):
    def __getstate__(self):
        return
    def __setstate__(self, state):
        return

aligner = PickablePairwiseAligner()
pickled_aligner = pickle.dumps(aligner)
unpickled_aligner = pickle.loads(pickled_aligner)
```

I do get only 1 alignment, while it should give plenty

```
alns = aligner.align('TTAGAGCACAGTTAG', 'TTACAGTTAACG')
len(alns)  # 14
alns = unpickled_aligner.align('TTAGAGCACAGTTAG', 'TTACAGTTAACG')
len(alns)  # 1
```

As I have some code processing these best scoring alignment to choose the one suiting my needs, the performance improvements reported earlier are related to this...

So in the end I cannot really tell whether pickling the aligner would indeed improve things.

**Comment by mdehoon** [comment]
Fixed by #3489
